File: backend/src/main/java/com/leadgrowth/controller/LeadController.java
package com.leadgrowth.controller;

import com.leadgrowth.dto.AddActivityLogRequest;
import com.leadgrowth.dto.CompleteStepRequest;
import com.leadgrowth.dto.ContactRepoDto;
import com.leadgrowth.dto.LeadDto;
import com.leadgrowth.dto.LeadNoteRequest;
import com.leadgrowth.dto.SalesActivityLogDto;
import com.leadgrowth.entity.LeadNote;
import com.leadgrowth.service.LeadService;
import jakarta.validation.Valid;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/leads")
public class LeadController {

    private final LeadService leadService;

    public LeadController(LeadService leadService) {
        this.leadService = leadService;
    }

    @GetMapping("/contacts")
    public ResponseEntity<List<ContactRepoDto>> getContactsRepository() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(leadService.getContactsRepository(email));
    }

    @GetMapping
    public ResponseEntity<List<LeadDto>> getLeads() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(leadService.getLeads(email));
    }

    @GetMapping("/high-priority")
    public ResponseEntity<List<LeadDto>> getHighPriorityLeads() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(leadService.getHighPriorityLeads(email));
    }

    @GetMapping("/new")
    public ResponseEntity<List<LeadDto>> getNewLeadsToday() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(leadService.getNewLeadsToday(email));
    }

    @GetMapping("/negotiation")
    public ResponseEntity<List<LeadDto>> getNegotiationLeads() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(leadService.getNegotiationLeads(email));
    }

    @PostMapping
    public ResponseEntity<LeadDto> createLead(@Valid @RequestBody LeadDto dto) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(leadService.createLead(dto, email));
    }

    @GetMapping("/{id}")
    public ResponseEntity<LeadDto> getLeadById(@PathVariable Long id) {
        return ResponseEntity.ok(leadService.getLeadById(id));
    }

    @PatchMapping("/{id}/status")
    public ResponseEntity<LeadDto> updateStatus(
            @PathVariable Long id,
            @RequestParam String status
    ) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(leadService.updateStatus(id, status, email));
    }

    @PatchMapping("/{id}/assign")
    @PreAuthorize("hasAnyRole('ADMIN', 'MANAGER', 'USER')")
    public ResponseEntity<LeadDto> assignLead(
            @PathVariable Long id,
            @RequestParam Long userId
    ) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(leadService.assignLead(id, userId, email));
    }

    @PostMapping("/{id}/notes")
    public ResponseEntity<Void> addNote(
            @PathVariable Long id,
            @Valid @RequestBody LeadNoteRequest request
    ) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        leadService.addNote(id, request, email);
        return ResponseEntity.ok().build();
    }

    @GetMapping("/{id}/notes")
    public ResponseEntity<List<LeadNote>> getNotes(@PathVariable Long id) {
        return ResponseEntity.ok(leadService.getNotes(id));
    }

    @PostMapping("/bulk-assign")
    @PreAuthorize("hasAnyRole('ADMIN', 'MANAGER')")
    public ResponseEntity<List<LeadDto>> bulkAssign(
            @RequestParam List<Long> leadIds,
            @RequestParam Long userId
    ) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(leadService.bulkAssignLeads(leadIds, userId, email));
    }

    @PostMapping("/bulk-random-assign")
    @PreAuthorize("hasAnyRole('ADMIN', 'MANAGER')")
    public ResponseEntity<List<LeadDto>> bulkRandomAssign(@RequestParam List<Long> leadIds) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(leadService.bulkRandomAssignLeads(leadIds, email));
    }

    @PostMapping("/bulk-status")
    public ResponseEntity<List<LeadDto>> bulkStatus(
            @RequestParam List<Long> leadIds,
            @RequestParam String status
    ) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(leadService.bulkUpdateLeadStatus(leadIds, status, email));
    }

    @PostMapping("/{id}/auto-assign")
    @PreAuthorize("hasAnyRole('ADMIN', 'MANAGER')")
    public ResponseEntity<LeadDto> autoAssignLead(@PathVariable Long id) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(leadService.assignLead(id, -1L, email));
    }

    @GetMapping("/workspace")
    public ResponseEntity<List<LeadDto>> getWorkspaceLeads() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(leadService.getLeads(email));
    }

    @PostMapping("/{id}/add-to-pipeline")
    @PreAuthorize("hasAnyRole('ADMIN', 'MANAGER', 'USER')")
    public ResponseEntity<LeadDto> addToPipeline(@PathVariable Long id) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(leadService.addToPipeline(id, email));
    }

    @GetMapping("/pipeline")
    @PreAuthorize("hasAnyRole('ADMIN', 'MANAGER', 'USER')")
    public ResponseEntity<List<LeadDto>> getPipelineLeads() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(leadService.getPipelineLeads(email));
    }

    @GetMapping("/pending-assigned")
    @PreAuthorize("hasAnyRole('ADMIN', 'MANAGER', 'USER')")
    public ResponseEntity<List<LeadDto>> getPendingAssignedLeads() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(leadService.getPendingAssignedLeads(email));
    }

    @PatchMapping("/{id}/activity")
    public ResponseEntity<LeadDto> updateActivity(
            @PathVariable Long id,
            @RequestParam String activityKey,
            @RequestParam String status,
            @RequestParam(required = false) String remarks
    ) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(leadService.updateLeadActivity(id, activityKey, status, remarks, email));
    }

    @PostMapping("/{id}/workflow-steps/{activityKey}/activities")
    public ResponseEntity<LeadDto> addStepActivityLog(
            @PathVariable Long id,
            @PathVariable String activityKey,
            @RequestBody AddActivityLogRequest request
    ) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(leadService.addStepActivityLog(id, activityKey, request, email));
    }

    @PostMapping("/{id}/workflow-steps/{activityKey}/complete")
    public ResponseEntity<LeadDto> completeWorkflowStep(
            @PathVariable Long id,
            @PathVariable String activityKey,
            @RequestBody(required = false) CompleteStepRequest request
    ) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(leadService.completeWorkflowStep(id, activityKey, request != null ? request : new CompleteStepRequest(), email));
    }

    @GetMapping("/{id}/activities-history")
    public ResponseEntity<List<SalesActivityLogDto>> getLeadActivityHistory(@PathVariable Long id) {
        return ResponseEntity.ok(leadService.getLeadActivityLogs(id));
    }

    @GetMapping("/workflow-pending-counts")
    public ResponseEntity<Map<String, Integer>> getWorkflowPendingCounts() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(leadService.getWorkflowPendingCounts(email));
    }

    @PatchMapping("/{id}/auto-save")
    public ResponseEntity<LeadDto> autoSaveWorkspaceLead(
            @PathVariable Long id,
            @RequestBody LeadDto dto
    ) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return ResponseEntity.ok(leadService.updateLeadWorkspace(id, dto, email));
    }
}

File: backend-dotnet/Controllers/LeadController.cs
using System.Security.Claims;
using LeadGrowth.DTOs;
using LeadGrowth.Models;
using LeadGrowth.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace LeadGrowth.Controllers;

[ApiController]
[Route("api/leads")]
[Authorize]
public class LeadController : ControllerBase
{
    private readonly ILeadService _leadService;

    public LeadController(ILeadService leadService)
    {
        _leadService = leadService;
    }

    [HttpGet("contacts")]
    public async Task<ActionResult<List<ContactRepoDto>>> GetContactsRepository()
    {
        var email = GetUserEmail();
        var contacts = await _leadService.GetContactsRepositoryAsync(email);
        return Ok(contacts);
    }

    [HttpGet]
    public async Task<ActionResult<List<LeadDto>>> GetLeads(
        [FromQuery] string? period,
        [FromQuery] string? startDate,
        [FromQuery] string? endDate)
    {
        var email = GetUserEmail();
        try
        {
            var leads = await _leadService.GetLeadsAsync(email, period, startDate, endDate);
            return Ok(leads);
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpGet("high-priority")]
    public async Task<ActionResult<List<LeadDto>>> GetHighPriorityLeads()
    {
        var email = GetUserEmail();
        var leads = await _leadService.GetHighPriorityLeadsAsync(email);
        return Ok(leads);
    }

    [HttpGet("new")]
    public async Task<ActionResult<List<LeadDto>>> GetNewLeadsToday()
    {
        var email = GetUserEmail();
        var leads = await _leadService.GetNewLeadsTodayAsync(email);
        return Ok(leads);
    }

    [HttpGet("negotiation")]
    public async Task<ActionResult<List<LeadDto>>> GetNegotiationLeads()
    {
        var email = GetUserEmail();
        var leads = await _leadService.GetNegotiationLeadsAsync(email);
        return Ok(leads);
    }

    [HttpPost]
    public async Task<ActionResult<LeadDto>> CreateLead([FromBody] LeadDto dto)
    {
        var email = GetUserEmail();
        try
        {
            var lead = await _leadService.CreateLeadAsync(dto, email);
            return Ok(lead);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpGet("{id}")]
    public async Task<ActionResult<LeadDto>> GetLeadById(long id)
    {
        try
        {
            var lead = await _leadService.GetLeadByIdAsync(id);
            return Ok(lead);
        }
        catch (ArgumentException ex)
        {
            return NotFound(new { message = ex.Message });
        }
    }

    [HttpPatch("{id}/status")]
    public async Task<ActionResult<LeadDto>> UpdateStatus(long id, [FromQuery] string status)
    {
        var email = GetUserEmail();
        try
        {
            var lead = await _leadService.UpdateStatusAsync(id, status, email);
            return Ok(lead);
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpPatch("{id}/assign")]
    [Authorize(Policy = "RequireUser")]
    public async Task<ActionResult<LeadDto>> AssignLead(long id, [FromQuery] long userId)
    {
        var email = GetUserEmail();
        try
        {
            var lead = await _leadService.AssignLeadAsync(id, userId, email);
            return Ok(lead);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpPost("{id}/notes")]
    public async Task<IActionResult> AddNote(long id, [FromBody] LeadNoteRequest request)
    {
        var email = GetUserEmail();
        try
        {
            await _leadService.AddNoteAsync(id, request, email);
            return Ok();
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpGet("{id}/notes")]
    public async Task<ActionResult<List<LeadNote>>> GetNotes(long id)
    {
        var notes = await _leadService.GetNotesAsync(id);
        return Ok(notes);
    }

    [HttpPost("bulk-assign")]
    [Authorize(Policy = "RequireManagerOrAdmin")]
    public async Task<ActionResult<List<LeadDto>>> BulkAssign([FromQuery] List<long> leadIds, [FromQuery] long userId)
    {
        var email = GetUserEmail();
        var leads = await _leadService.BulkAssignLeadsAsync(leadIds, userId, email);
        return Ok(leads);
    }

    [HttpPost("bulk-random-assign")]
    [Authorize(Policy = "RequireManagerOrAdmin")]
    public async Task<ActionResult<List<LeadDto>>> BulkRandomAssign([FromQuery] List<long> leadIds)
    {
        var email = GetUserEmail();
        var leads = await _leadService.BulkRandomAssignLeadsAsync(leadIds, email);
        return Ok(leads);
    }

    [HttpPost("bulk-status")]
    public async Task<ActionResult<List<LeadDto>>> BulkStatus([FromQuery] List<long> leadIds, [FromQuery] string status)
    {
        var email = GetUserEmail();
        var leads = await _leadService.BulkUpdateLeadStatusAsync(leadIds, status, email);
        return Ok(leads);
    }

    [HttpPost("{id}/auto-assign")]
    [Authorize(Policy = "RequireManagerOrAdmin")]
    public async Task<ActionResult<LeadDto>> AutoAssignLead(long id)
    {
        var email = GetUserEmail();
        try
        {
            var lead = await _leadService.AssignLeadAsync(id, -1, email);
            return Ok(lead);
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpGet("workspace")]
    public async Task<ActionResult<List<LeadDto>>> GetWorkspaceLeads()
    {
        var email = GetUserEmail();
        try
        {
            var leads = await _leadService.GetLeadsAsync(email);
            return Ok(leads);
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpPost("{id}/add-to-pipeline")]
    [Authorize(Policy = "RequireUser")]
    public async Task<ActionResult<LeadDto>> AddToPipeline(long id)
    {
        var email = GetUserEmail();
        try
        {
            var lead = await _leadService.AddToPipelineAsync(id, email);
            return Ok(lead);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpGet("pipeline")]
    [Authorize(Policy = "RequireUser")]
    public async Task<ActionResult<List<LeadDto>>> GetPipelineLeads()
    {
        var email = GetUserEmail();
        try
        {
            var leads = await _leadService.GetPipelineLeadsAsync(email);
            return Ok(leads);
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpGet("pending-assigned")]
    [Authorize(Policy = "RequireUser")]
    public async Task<ActionResult<List<LeadDto>>> GetPendingAssignedLeads()
    {
        var email = GetUserEmail();
        var leads = await _leadService.GetPendingAssignedLeadsAsync(email);
        return Ok(leads);
    }

    [HttpPatch("{id}/activity")]
    public async Task<ActionResult<LeadDto>> UpdateActivity(long id, [FromQuery] string activityKey, [FromQuery] string status, [FromQuery] string? remarks)
    {
        var email = GetUserEmail();
        try
        {
            var lead = await _leadService.UpdateLeadActivityAsync(id, activityKey, status, remarks, email);
            return Ok(lead);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpPost("{id}/workflow-steps/{activityKey}/activities")]
    public async Task<ActionResult<LeadDto>> AddStepActivityLog(long id, string activityKey, [FromBody] AddActivityLogRequest request)
    {
        var email = GetUserEmail();
        try
        {
            var lead = await _leadService.AddStepActivityLogAsync(id, activityKey, request, email);
            return Ok(lead);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpPost("{id}/workflow-steps/{activityKey}/complete")]
    public async Task<ActionResult<LeadDto>> CompleteWorkflowStep(long id, string activityKey, [FromBody] CompleteStepRequest? request)
    {
        var email = GetUserEmail();
        try
        {
            var lead = await _leadService.CompleteWorkflowStepAsync(id, activityKey, request, email);
            return Ok(lead);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpGet("{id}/activities-history")]
    public async Task<ActionResult<List<SalesActivityLogDto>>> GetLeadActivityHistory(long id)
    {
        var logs = await _leadService.GetLeadActivityLogsAsync(id);
        return Ok(logs);
    }

    [HttpGet("workflow-pending-counts")]
    public async Task<ActionResult<Dictionary<string, int>>> GetWorkflowPendingCounts()
    {
        var email = GetUserEmail();
        var counts = await _leadService.GetWorkflowPendingCountsAsync(email);
        return Ok(counts);
    }

    [HttpPatch("{id}/auto-save")]
    public async Task<ActionResult<LeadDto>> AutoSaveWorkspaceLead(long id, [FromBody] LeadDto dto)
    {
        var email = GetUserEmail();
        try
        {
            var lead = await _leadService.UpdateLeadWorkspaceAsync(id, dto, email);
            return Ok(lead);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { message = ex.Message });
        }
    }

    [HttpGet("{id}/timeline")]
    public async Task<ActionResult<List<LeadHistoryDto>>> GetLeadTimeline(long id)
    {
        var timeline = await _leadService.GetLeadTimelineAsync(id);
        return Ok(timeline);
    }

    private string GetUserEmail()
    {
        return User.FindFirstValue(ClaimTypes.Name) ?? User.FindFirstValue(ClaimTypes.Email) ?? string.Empty;
    }
}
